proj/hps_accel/gateware/stream/stream.py

```diff
@@ -14,146 +14,106 @@
 # limitations under the License.
 
 from nmigen import Shape
-from nmigen.hdl.rec import Record, DIR_FANIN, DIR_FANOUT
+from nmigen.hdl.rec import Layout, Record
 
+__all__ = ['StreamDefinition', 'Stream', 'connect']
 
-class _Endpoint:
-    """Abstract base class for Sinks and Sources."""
 
-    def __init__(self, payload_type, name, src_loc_at):
-        self.payload_type = payload_type
-        self._record = Record([
-            ("valid", Shape(), DIR_FANOUT),
-            ("ready", Shape(), DIR_FANIN),
-            ("last", Shape(), DIR_FANOUT),
-            ("payload", payload_type, DIR_FANOUT),
-        ], src_loc_at=2+src_loc_at, name=name)
-        self.valid = self._record.valid
-        self.ready = self._record.ready
-        self.last = self._record.last
-        self.payload = self._record.payload
+class StreamDefinition:
+    """Defines a stream and the guarantees that it makes.
 
-    def is_transferring(self):
-        """Returns an expression that is true when a transfer takes place."""
-        return (self.valid & self.ready)
+    ignores_valid:
+      Stream may use payload without valid being set.
 
-
-class Source(_Endpoint):
-    """A stream source.
-
-    Parameters
-    ----------
-
-    payload_type: Shape(N) or Layout
-      The payload transferred from this Source.
-    name: str
-      Base for signal names.
-
-    Attributes:
-    -----------
-
-    payload_type: Shape(N) or Layout
-    valid: Signal(1), out
-    ready: Signal(1), in
-    last: Signal(1), out
-    payload: Signal(N) or Record, out
+    ignores_ready:
+      Stream may present a new payload while valid is asserted
+      and before ready is asserted.
     """
 
-    def __init__(self, payload_type, name=None, src_loc_at=0):
-        super().__init__(payload_type, name, src_loc_at)
-
-    def connect(self, sink):
-        """Returns a list of statements that connects this source to a sink.
-
-        Parameters:
-          sink: This Sink to which to connect.
-        """
-        assert isinstance(sink, Sink)
-        return self._record.connect(sink._record)
-
-
-class Sink(_Endpoint):
-    """A stream sink
-
+    @staticmethod
+    def cast(obj, src_loc_at=0, ignores_ready=False, ignores_valid=False):
+        if isinstance(obj, StreamDefinition):
+            return StreamDefinition(
+                paylad_type=obj.payload_type,
```

Comment:
paylad -> payload

Reply:
Hmm. More test cases needed, clearly.

Reply:
The new version of this function is used in the normal path, so this is now tested - at least cursorily tested.